### PR TITLE
[SPARK-13540][SQL] Supports using nested classes within Scala objects as Dataset element type

### PR DESCRIPTION
**Pull request by liancheng**
## What changes were proposed in this pull request?

Nested classes defined within Scala objects are translated into Java static nested classes. Unlike inner classes, they don't need outer scopes. But the analyzer still thinks that an outer scope is required.

This PR fixes this issue simply by checking whether a nested class is static before looking up its outer scope.
## How was this patch tested?

A test case is added to `DatasetSuite`. It checks contents of a Dataset whose element type is a nested class declared in a Scala object.

### CODE REVIEW

**Comment by liancheng** [comment]
cc @cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 1]
```diff
@@ -559,7 +561,10 @@ class Analyzer(
       }
 
       resolveExpression(unbound, LocalRelation(attributes), throws = true) transform {
-        case n: NewInstance if n.outerPointer.isEmpty && n.cls.isMemberClass =>
+        case n: NewInstance
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: would be great to add some inline documentation what about these conditions are checking

File: sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala [hunk 2]
```diff
@@ -621,12 +621,23 @@ class DatasetSuite extends QueryTest with SharedSQLContext {
       ds.filter(_ => true),
       Some(1), Some(2), Some(3))
   }
+
+  test("SPARK-13540 add Scala object to OuterScopes") {
+    OuterScopes.addOuterScope(OuterObject)
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala]: I'm a little confused here. If nested classes defined in scala object don't need out scope, why do we add out scope in this test case?

Comment (liancheng) [inline sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala]: Oh, at first I got a wrong fix for this issue (see the first commit).  I thought the problem was that when adding the Scala object into `OuterScopes`, the class name without the `$` suffix should be added. Later I found I was wrong, but forgot to remove this line. Thanks!

**Comment by liancheng** [comment]
Thanks for the review! Comments addressed.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 3]
```diff
@@ -559,7 +561,13 @@ class Analyzer(
...
+          // If this is an inner class of another class, register the outer object in `OuterScopes`.
+          // Note that static inner classes (e.g., inner classes within Scala objects. don't need
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: nit: `(e.g. inner classes within Scala objects)`

File: sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala [hunk 4]
```diff
@@ -621,12 +621,22 @@ class DatasetSuite extends QueryTest with SharedSQLContext {
       ds.filter(_ => true),
       Some(1), Some(2), Some(3))
   }
+
+  test("SPARK-13540 add Scala object to OuterScopes") {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala]: the test name should be updated.

**Comment by cloud-fan** [comment]
LGTM except some minor comments.

**Comment by liancheng** [comment]
retest this please

**Comment by cloud-fan** [comment]
retest this please.

**Comment by liancheng** [comment]
I'm merging this to master as it blocks #11431.

**Pull request closed without merging**